### ISSUE TITLE
DynamicPaddedBatchDatasetOp for tf.data

### ISSUE THREAD

**Issue by feihugis** [tensorflow/tensorflow#39924]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): tf-nightly
- Are you willing to contribute it (Yes/No): Yes. If this feature is agreed, I'm happy to implement it.

**Describe the feature and the current behavior/state.**

`tf.data` provides a `padded_batch` transformation, but it only supports a fixed batch size. In some use cases (e.g. model inference), `dynamic_padded_batch` will bring some extra performance benefits. 

For example, for the named entity recognition model (e.g. BiLSTM-based, BERT-based), the input sentences/tokens will be required to be padded and then batched before feed into the model inference. If we use the fixed batch size, the short sentences and the long sentences may be combined together in a batch, which will add many unnecessary padding elements to the short sentences. For this kind of cases, `dynamic_padded_batch` transformation will be more efficient. It enables users to input a customized `dynamic_batch_function` to control how to split the batches. In this case, the sentences with similar lengths may be put together as a batch; If the next sentence is too long, it will be placed into the next batch; With this, the number of padding elements in each batch can be as small as possible. 

In summary,  `DynamicPaddedBatchDatasetOp` will provide users with more flexibility to control how to split the batch.

**Will this change the current api? How?**

It will not change the current api; another new operation `DynamicPaddedBatchDatasetOp` will be added; 

**Who will benefit with this feature?**

The users who use `tf.data` to build the data pipeline.

**Any Other info.**

cc: @jsimsa @aaudiber

**Comment by jsimsa** [comment]
I think this use case could be addressed using the existing `tf.data.experimental.group_by_window` transformation. For instance, for the input pipeline for the official Transformer model uses this transformation to group sentences based on length and creates padded batches for each bucket separately.

**Comment by feihugis** [comment]
Thanks very much @jsimsa! Yeah, `group_by_window` can be used here. The example in the official Transformer model is exactly what I want do. Thanks again!

**Comment by feihugis** [comment]
Closed this issue as it has been resolved.
